src/uproot/source/fsspec.py

```diff
@@ -0,0 +1,128 @@
+import uproot.source.chunk
+
+
+class FSSpecSource(uproot.source.chunk.Source):
+    """
+    Args:
+        file_path (str): A URL for the file to open.
+        **kwargs (dict): any extra arguments to be forwarded to the particular
+            FileSystem instance constructor. This might include S3 access keys,
+            or HTTP headers, etc.
+
+    A :doc:`uproot.source.chunk.Source` that uses FSSpec's cat_ranges feature
+    to get many chunks in one request.
+    """
+
+    def __init__(self, file_path, **kwargs):
+        import fsspec.core
+
+        # Remove uproot-specific options (should be done earlier)
+        # TODO: is timeout always valid?
+        opts = {
+            k: v
+            for k, v in kwargs.items()
+            if k
+            not in {
+                "file_handler",
+                "xrootd_handler",
+                "http_handler",
+                "object_handler",
+                "max_num_elements",
+                "num_workers",
+                "num_fallback_workers",
+                "begin_chunk_size",
+                "minimal_ttree_metadata",
+            }
+        }
```

Comment:
I would propose to pop uproot-specific options in `uproot.reading.open` instead of passing all of them to the source. I think it would also be nice to deprecate specific `*_handler` functions in favor of some sort of `filesystem=` argument which would override the default irrespective of what the default is. Some of these arguments are source-related (e.g. `timeout` or `num_workers`), in which case they would better be defaulted keyword arguments in the respective source constructors.

Reply:
The `*_handler` arguments are wrong for a Source like fsspec, I agree. (And didn't see it coming.)

If we go the route of handling all remote file I/O with fsspec, then the natural thing to do would be to replace these arguments with a `filesystem` or `fs` argument (as is becoming standard in some other projects).

The role currently played by `file_handler` and `object_handler` would need some equivalent, though. There are two ways one might want to use a `file_handler`: as memmap and as conventional file object(s). I expect that memmap would usually be better, but worse through NFS. For `object_handler`, there's only one option (not something for a user to select), but it is very distinct from handlers for an object that you know is a file: if it's a generic object, we can't assume that we can `seek` and `read` in parallel.